alerting/src/main/kotlin/org/opensearch/alerting/model/action/ActionExecutionScope.kt

```diff
@@ -0,0 +1,179 @@
+import org.opensearch.common.io.stream.StreamOutput
+import org.opensearch.common.io.stream.Writeable
+import org.opensearch.common.xcontent.ToXContent
+import org.opensearch.common.xcontent.ToXContentObject
+import org.opensearch.common.xcontent.XContentBuilder
+import org.opensearch.common.xcontent.XContentParser
+import org.opensearch.common.xcontent.XContentParser.Token
+import org.opensearch.common.xcontent.XContentParserUtils.ensureExpectedToken
+import java.io.IOException
+import java.lang.IllegalArgumentException
+
+/**
+ * This class represents configurations used to control the scope of Action executions when Alerts are created.
+ */
+sealed class ActionExecutionScope : Writeable, ToXContentObject {
+
+    enum class Type { PER_ALERT, PER_EXECUTION }
+
+    companion object {
+        const val PER_ALERT_FIELD = "per_alert"
+        const val PER_EXECUTION_FIELD = "per_execution"
+        const val ACTIONABLE_ALERTS_FIELD = "actionable_alerts"
+
+        @JvmStatic
+        @Throws(IOException::class)
+        fun parse(xcp: XContentParser): ActionExecutionScope {
+            var type: Type? = null
+            var actionExecutionScope: ActionExecutionScope? = null
+            val alertFilter = mutableSetOf<AlertCategory>()
+
+            ensureExpectedToken(Token.START_OBJECT, xcp.currentToken(), xcp)
+            while (xcp.nextToken() != Token.END_OBJECT) {
+                val fieldName = xcp.currentName()
+                xcp.nextToken()
+
+                // If the type field has already been set, the user has provided more than one type of schedule
+                if (type != null) {
+                    throw IllegalArgumentException("You can only specify one type of action execution scope.")
+                }
+
+                when (fieldName) {
+                    PER_ALERT_FIELD -> {
+                        type = Type.PER_ALERT
+                        while (xcp.nextToken() != Token.END_OBJECT) {
+                            val perAlertFieldName = xcp.currentName()
+                            xcp.nextToken()
+                            when (perAlertFieldName) {
+                                ACTIONABLE_ALERTS_FIELD -> {
+                                    ensureExpectedToken(Token.START_ARRAY, xcp.currentToken(), xcp)
+                                    val allowedCategories = AlertCategory.values().map { it.toString() }
+                                    while (xcp.nextToken() != Token.END_ARRAY) {
+                                        val alertCategory = xcp.text()
+                                        if (!allowedCategories.contains(alertCategory)) {
+                                            throw IllegalStateException("Actionable alerts should be one of $allowedCategories")
+                                        }
+                                        alertFilter.add(AlertCategory.valueOf(alertCategory))
+                                    }
+                                }
```

Comment:
What's the expectation in else block here? Shouldn't it fail if `perAlertFieldName` is not `ACTIONABLE_ALERTS_FIELD`?

Reply:
Yeah, I'll add an `else` to make the `when` exhaustive.